src/tools.cpp

```diff
@@ -155,30 +156,83 @@ inline bool seemsToBeHtml(const std::string& path)
   return false;
 }
 
+static std::string maybeResolveLink(const std::string& path) {
+  // TODO: The implementation can be replaced with the following snippet when
+  //       compiler support for <filesystem> is sufficient enough.
+  // std::filesystem::path p(path);
+  // if(std::filesystem::exists(p) && std::filesystem::is_symlink(p)) {
+  //   return std::filesystem::read_symlink(p);
+  // } else {
+  //   return path;
+  // }
+  for (int retry = 0; retry < 5; ++retry) {
+    struct stat path_stat;
+    if (lstat(path.c_str(), &path_stat) == -1) {
+      std::cerr << "zimwriterfs: unable to stat file at path: " << path
+                << std::endl;
+      throw(errno);
+    }
+    if (S_ISREG(path_stat.st_mode)) {
+      return path;
+    }
+    if (!S_ISLNK(path_stat.st_mode)) {
+      std::cerr << "zimwriterfs: file is neither a regular file nor a symbolic "
+                << "link: " << path << std::endl;
+      return path;
+    }
+    std::string resolved_path;
+    resolved_path.resize(path_stat.st_size + 1);
+    // Allow readlink to write st_size + 1 bytes even though we will truncate
+    // resolved_path to st_size. This way, we can detect if the target path
+    // changed inbetween calls to lstat and readlink because we expect only
+    // st_size of bytes to be written to the buffer.
+    // The only case where link_len > st_size is true is when the target
+    // path has changed.
+    ssize_t link_len = readlink(
+        path.c_str(), &resolved_path[0], path_stat.st_size + 1);
+    if (link_len < 0) {
+      std::cerr << "zimwriterfs: unable to resolve symlink: " << path
+                << std::endl;
+      throw(errno);
+    }
+    if (link_len > path_stat.st_size) {
+      std::cerr << "zimwriterfs: symbolic link target path increased in size "
+                << "between lstat and readlink: " << path << std::endl;
+      continue;
+    }
+    resolved_path.resize(path_stat.st_size);
```

Comment:
We should resize to `link_len`. If the target path has changed between `lstat` and `readlink`, the correct value is `link_len`, not `path_stat.st_size`.